### PR TITLE
Improvements to createReaderGroup and updateReaderGroup APIs

### LINKED ISSUE #5447: UpdateReaderGroup API should return generation of the ReaderGroupConfig post update
**Problem description**
UpdateReaderGroup API should return generation of the RGConfig post update

**Problem location**
Controller

**Suggestions for an improvement**
UpdateReaderGroup API should return generation of the RGConfig post update
Currently it only returns a status indicating if the update succeeded or failed.

### PR DESCRIPTION
**Pull request by pbelgundi**
**Change log description**  
This PR has the following changes:
1. createReaderGroup() method in streamMetadataTasks has been changed to do only 1 metadata operation instead of multiple. The the rest of the create operations have been moved to the CreateReaderGroupTask.
2. createReaderGroup() for internal reader groups created by ControllerEventProcessor should not use events as it leads to cyclic dependency, so changed LocalController::createReaderGroup() to invoke a createReaderGroupInternal() method on StreamMetadataTasks for creating a reader group without using events.
3. Changed updateReaderGroup API to return updated generation of the ReaderGroup in addition to update operation status.

**Purpose of the change**  
Fixes #5447

**What the code does**  

**How to verify it**  
All Unit and Integration tests should pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java [hunk 1]
```diff
@@ -1571,8 +1571,8 @@ private void closeChannel() {
         final long requestId = requestIdGenerator.get();
         long traceId = LoggerHelpers.traceEnter(log, "updateReaderGroup", rgConfig, requestId);
 
-        final CompletableFuture<UpdateReaderGroupStatus> result = this.retryConfig.runAsync(() -> {
-            RPCAsyncCallback<UpdateReaderGroupStatus> callback = new RPCAsyncCallback<>(requestId, "updateReaderGroup", scope, rgName, rgConfig);
+        final CompletableFuture<UpdateReaderGroupResponse> result = this.retryConfig.runAsync(() -> {
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java:1574]: why are we changing the name of this class `UpdateReaderGroupStatus`?

File: client/src/main/java/io/pravega/client/control/impl/ModelHelper.java [hunk 2]
```diff
@@ -271,10 +271,10 @@ public static ReaderGroupConfig encode(Controller.ReaderGroupConfiguration rgCon
                 .generation(rgConfig.getGeneration())
                 .readerGroupId(UUID.fromString(rgConfig.getReaderGroupId()))
                 .startingStreamCuts(rgConfig.getStartingStreamCutsList().stream()
-                        .collect(Collectors.toMap(streamCut -> Stream.of(streamCut.getStreamInfo().getScope(),
-                                streamCut.getStreamInfo().getStream()),
-                                streamCut -> new StreamCutImpl(Stream.of(streamCut.getStreamInfo().getScope(), streamCut.getStreamInfo().getStream()),
-                                        getSegmentOffsetMap(streamCut.getStreamInfo().getScope(), streamCut.getStreamInfo().getStream(), streamCut.getCutMap())))))
+                .collect(Collectors.toMap(streamCut -> Stream.of(streamCut.getStreamInfo().getScope(),
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/control/impl/ModelHelper.java:274]: nit: please revert this

File: client/src/main/java/io/pravega/client/control/impl/ModelHelper.java [hunk 3]
```diff
@@ -283,7 +283,7 @@ public static ReaderGroupConfig encode(Controller.ReaderGroupConfiguration rgCon
                 .build();
     }
 
-    private static Map<Segment, Long> getSegmentOffsetMap(String streamName, String scopeName, Map<Long, Long> streamCutMap) {
+    public static Map<Segment, Long> getSegmentOffsetMap(String scopeName, String streamName, Map<Long, Long> streamCutMap) {
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/control/impl/ModelHelper.java:286]: good catch.

File: client/src/main/java/io/pravega/client/control/impl/ModelHelper.java [hunk 4]
```diff
@@ -464,12 +464,12 @@ public static final NodeUri decode(final PravegaNodeUri uri) {
         Preconditions.checkNotNull(groupName, "ReaderGroup name is null");
         Preconditions.checkNotNull(config, "ReaderGroupConfig is null");
 
-        List<StreamCut> startStreamCuts = config.getStartingStreamCuts().entrySet().stream()
+        List<Controller.StreamCut> startStreamCuts = config.getStartingStreamCuts().entrySet().stream()
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/control/impl/ModelHelper.java:467]: nit: revert this

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java [hunk 5]
```diff
@@ -201,11 +215,75 @@ public LocalController(ControllerService controller, boolean authorizationEnable
...
+                    if (!exists) {
+                            return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SCOPE_NOT_FOUND);
+                    }
+                    //2. check state of the ReaderGroup
+                    return Futures.exceptionallyExpecting(streamMetadataStore.getReaderGroupState(scope, rgName, true, null, localExecutor),
+                           e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, ReaderGroupState.UNKNOWN)
+                           .thenCompose(state -> {
+                           if (state.equals(ReaderGroupState.UNKNOWN) || state.equals(ReaderGroupState.CREATING)) {
+                              return streamMetadataStore.addReaderGroupToScope(scope, rgName, config.getReaderGroupId())
+                              .thenCompose(v -> streamMetadataStore.createReaderGroup(scope, rgName, config, createTimestamp, context, localExecutor)
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java:249]: i suggest we move all of the code below this to a common helper method in streamMetadataTasks which is invoked both from here and from CreateRGTask

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java [hunk 6]
```diff
@@ -57,48 +65,78 @@ public CreateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
...
+                        log.warn("Skipping processing of CreateReaderGroupEvent with invalid UUID.");
+                        return CompletableFuture.completedFuture(null);
                     }
-                    return CompletableFuture.completedFuture(null);
+                    ReaderGroupConfig config = getConfigFromEvent(request);
+                    return Futures.exceptionallyExpecting(streamMetadataStore.getReaderGroupState(scope, readerGroup, true, null, executor),
+                            e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, ReaderGroupState.UNKNOWN)
+                            .thenCompose(state -> {
+                                if (state.equals(ReaderGroupState.UNKNOWN) || state.equals(ReaderGroupState.CREATING)) {
+                                    return streamMetadataStore.createReaderGroup(scope, readerGroup, config, System.currentTimeMillis(), context, executor)
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java:83]: this code is very similar to the one in local controller.. lets consolidate it at a single place please.

**Review by shiveshr — CHANGES_REQUESTED** [review]
Looking good overall except few suggestions

File: client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java (hunk 1 above)
Comment (pbelgundi) [reply client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java:1574]: UpdateReaderGroupResponse has 'Status' of the operation and 'generation' post upgrade. Hence the change in name.

File: client/src/main/java/io/pravega/client/control/impl/ModelHelper.java (hunk 4 above)
Comment (pbelgundi) [reply client/src/main/java/io/pravega/client/control/impl/ModelHelper.java:467]: This is necessary because this class deals with 2 kind of StreamCut class objects - one from Controller and another one from Client.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java (hunk 5 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java:249]: done

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java (hunk 6 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java:83]: done

File: client/src/main/java/io/pravega/client/control/impl/ModelHelper.java (hunk 2 above)
Comment (pbelgundi) [reply client/src/main/java/io/pravega/client/control/impl/ModelHelper.java]: ok done.

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStreamMetadataStore.java [hunk 7]
```diff
@@ -307,20 +307,9 @@ PravegaTablesReaderGroup newReaderGroup(final String scope, final String rgName)
...
-                                                                 final RGOperationContext context,
-                                                                 final Executor executor) {
-        return Futures.completeOn(((PravegaTablesScope) getScope(scope)).addReaderGroupToScope(name, configuration.getReaderGroupId())
-                .thenCompose(rgAdded -> {
-                   if (rgAdded) {
-                       return super.createReaderGroup(scope, name, configuration, createTimestamp, context, executor);
-                   }
-                   return CompletableFuture.completedFuture(null);
-                }), executor);
+    public CompletableFuture<Void> addReaderGroupToScope(final String scope,
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStreamMetadataStore.java:310]: nit: override annotation

File: controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java [hunk 8]
```diff
@@ -52,10 +53,10 @@
     public static ReaderGroupConfigRecord update(ReaderGroupConfig rgConfig, long generation, boolean isUpdating) {
         Map<String, RGStreamCutRecord> startStreamCuts = rgConfig.getStartingStreamCuts().entrySet().stream()
                 .collect(Collectors.toMap(e -> e.getKey().getScopedName(),
-                        e -> new RGStreamCutRecord(getStreamCutMap(e.getValue()))));
+                        e -> new RGStreamCutRecord(ImmutableMap.copyOf(ModelHelper.getStreamCutMap(e.getValue())))));
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java:56]: can we make modelhelper return an imutable map itself. rather than create another copy here.

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java [hunk 9]
```diff
@@ -331,24 +333,113 @@ private ReaderGroupConfiguration getRGConfigurationFromRecord(final String scope
...
-                }
-                // idempotent call
-                return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SUCCESS);
-             });
+         return isRGCreationComplete(scope, rgName)
+                 .thenCompose(complete -> {
+                     if (!complete) {
+                         Map<String, RGStreamCutRecord> startStreamCuts = config.getStartingStreamCuts().entrySet().stream()
+                                 .collect(Collectors.toMap(e -> e.getKey().getScopedName(),
+                                         e -> new RGStreamCutRecord(ImmutableMap.copyOf(ModelHelper.getStreamCutMap(e.getValue())))));
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:341]: same here and everywhere else where this method is called. we are creating copies that could be avoided if the method returned an immutable map itself

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java [hunk 10]
```diff
@@ -331,24 +333,113 @@ private ReaderGroupConfiguration getRGConfigurationFromRecord(final String scope
...
+                                .thenCompose(complete -> {
+                                    if (!complete) {
+                                        return streamMetadataStore.addReaderGroupToScope(scope, rgName, config.getReaderGroupId())
+                                                .thenCompose(x -> createReaderGroupTasks(scope, rgName, config, System.currentTimeMillis()))
+                                                .thenApply(y -> CreateReaderGroupStatus.Status.SUCCESS);
+                                    }
+                                    return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SUCCESS);
+                                });
+                    });
+        }, e -> Exceptions.unwrap(e) instanceof RetryableException, 10, executor);
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:440]: we should use RetryableException.isRetryable(e) predicate.

**Review by shiveshr — CHANGES_REQUESTED** [review]
LGTM, some minor improvements.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java [hunk 11]
```diff
@@ -57,48 +57,48 @@ public CreateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
...
-                                             readerGroup, status.toString())));
-                           })).exceptionally(ex -> {
-                                    log.debug(ex.getMessage());
-                                    Throwable cause = Exceptions.unwrap(ex);
-                                    throw new CompletionException(cause);
-                           });
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getReaderGroupId(scope, readerGroup, context, executor)
+                .thenCompose(rgId -> {
+                    if (!rgId.equals(readerGroupId)) {
+                        log.warn("Skipping processing of CreateReaderGroupEvent with invalid UUID.");
```
[the remark this answers is not in the document]
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java:67]: Ok.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java [hunk 12]
```diff
@@ -57,48 +57,48 @@ public CreateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
...
+                        return CompletableFuture.completedFuture(null);
                     }
-                    return CompletableFuture.completedFuture(null);
+                    ReaderGroupConfig config = getConfigFromEvent(request);
```
[the remark this answers is not in the document]
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java:70]: done

File: controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java (hunk 8 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java:56]: Another method that uses getStreamCutMap requires a Map and not an Immutable map.
Given that size of this map should be pretty small typically < 10 it should be fine to do the conversion.

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java (hunk 9 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:341]: Same as earlier.

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java (hunk 10 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:440]: Can be changed in another PR where we change this for all code using RetryableException

File: controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStreamMetadataStore.java (hunk 7 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStreamMetadataStore.java:310]: done

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/StreamRequestHandler.java [hunk 13]
```diff
@@ -119,8 +119,8 @@ public StreamRequestHandler(AutoScaleTask autoScaleTask,
 
     @Override
     public CompletableFuture<Void> processCreateReaderGroup(CreateReaderGroupEvent createRGEvent) {
-        log.info("Processing create request {} for ReaderGroup {}/{}",
-                createRGEvent.getRequestId(), createRGEvent.getScopeName(), createRGEvent.getRgName());
+        log.info("Processing create request for ReaderGroup {}/{}",
```
[the remark this answers is not in the document]
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/StreamRequestHandler.java:122]: ok done.

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java [hunk 14]
```diff
@@ -57,48 +57,48 @@ public CreateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
...
-                                             readerGroup, status.toString())));
-                           })).exceptionally(ex -> {
-                                    log.debug(ex.getMessage());
-                                    Throwable cause = Exceptions.unwrap(ex);
-                                    throw new CompletionException(cause);
-                           });
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getReaderGroupId(scope, readerGroup, context, executor)
+                .thenCompose(rgId -> {
+                    if (!rgId.equals(readerGroupId)) {
+                        log.warn("Skipping processing of CreateReaderGroupEvent with stale UUID.");
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java:68]: also this doesnt need to be warn, does it? its idempotent no op case. so we should add it at info and not warn

File: test/integration/src/test/java/io/pravega/test/integration/controller/server/ControllerServiceTest.java [hunk 15]
```diff
@@ -135,8 +135,7 @@ public void streamMetadataTest() throws Exception {
         assertFalse(controller.createStream(scope, stream, streamConfiguration).join());
     }
     
-    
-    @Test(timeout = 80000)
+    @Test(timeout = 30000)
```
Comment (shiveshr) [inline test/integration/src/test/java/io/pravega/test/integration/controller/server/ControllerServiceTest.java:138]: please revert this change

File: shared/controller-api/src/main/java/io/pravega/shared/controller/event/CreateReaderGroupEvent.java [hunk 16]
```diff
@@ -61,16 +73,38 @@ protected void declareVersions() {
...
+            ImmutableMap.Builder<String, RGStreamCutRecord> startStreamCutBuilder = ImmutableMap.builder();
+            source.readMap(DataInput::readUTF, RGStreamCutRecord.SERIALIZER::deserialize, startStreamCutBuilder);
+            eb.startingStreamCuts(startStreamCutBuilder.build());
+            ImmutableMap.Builder<String, RGStreamCutRecord> endStreamCutBuilder = ImmutableMap.builder();
+            source.readMap(DataInput::readUTF, RGStreamCutRecord.SERIALIZER::deserialize, endStreamCutBuilder);
+            eb.endingStreamCuts(endStreamCutBuilder.build());
         }
     }
     //endregion
+
```
Comment (shiveshr) [inline shared/controller-api/src/main/java/io/pravega/shared/controller/event/CreateReaderGroupEvent.java:108]: nit: please remove the extra line

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java (hunk 14 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java:68]: Its an unexpected thing to happen hence added it as warn

File: controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java (hunk 8 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java:56]: ImmutableMap extends map. so in model helper you can construct an immutable map itself.

**Review by shiveshr — APPROVED** [review]
LGTM, couple of minor comments. i am ok if you address them in a separate PR too.

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java (hunk 10 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:440]: please open a github issue for this. thanks

File: shared/controller-api/src/main/java/io/pravega/shared/controller/event/CreateReaderGroupEvent.java (hunk 16 above)
Comment (pbelgundi) [reply shared/controller-api/src/main/java/io/pravega/shared/controller/event/CreateReaderGroupEvent.java:108]: done in PR 5466

File: test/integration/src/test/java/io/pravega/test/integration/controller/server/ControllerServiceTest.java (hunk 15 above)
Comment (pbelgundi) [reply test/integration/src/test/java/io/pravega/test/integration/controller/server/ControllerServiceTest.java:138]: done in PR5466

File: controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java (hunk 8 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/store/stream/records/ReaderGroupConfigRecord.java:56]: done in PR5466

File: controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java (hunk 9 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java:341]: done in PR5466
